streams/shared/src/main/scala/zio/stream/ZTransducer.scala

```diff
@@ -98,6 +98,95 @@ abstract class ZTransducer[-R, +E, -I, +O](val push: ZManaged[R, Nothing, Option
    */
   final def mapM[R1 <: R, E1 >: E, P](f: O => ZIO[R1, E1, P]): ZTransducer[R1, E1, I, P] =
     ZTransducer[R1, E1, I, P](self.push.map(push => i => push(i).flatMap(_.mapM(f))))
+
+  /**
+   * Runs both transducers in parallel on the input, emitting the result from the
+   * one that finishes first.
+   * If both finished simultaneously, the first one has priority.
+   *
+   * NB! Result might depend on the way how the stream is chunked.
+   * See {@link #raceBoth} for details.
+   */
+  def race[R1 <: R, E1 >: E, I1 <: I, O1 >: O](
+    that: ZTransducer[R1, E1, I1, O1]
+  ): ZTransducer[R1, E1, I1, O1] =
+    self.raceBoth(that).map(_.merge)
+
+  /**
+   * Runs both transducers in parallel on the input, emitting the result from the
+   * one that finishes first.
+   * If both finished simultaneously, the first one has priority.
+   *
+   * NB! Due to the batch nature of streams,
+   * result might depend on the way how the stream is chunked.
+   * If one of the transducers emitted a value at the moment `t1`,
+   * and then another emitted a value at the moment `t2 > t1`,
+   * the second value could be emitted instead of the first
+   * in case `t1` and `t2` belong to the same chunk.
+   *
+   * If such deviations are not tolerated,
+   * consider rechunking the stream before transducing:
+   * `stream.chunkN(1).transduce(a.raceBoth(b))`
+   */
+  def raceBoth[R1 <: R, E1 >: E, I1 <: I, O1](
+    that: ZTransducer[R1, E1, I1, O1]
+  ): ZTransducer[R1, E1, I1, Either[O, O1]] = {
+    case class State(leftDebt: Long, rightDebt: Long)
+
+    def updateState(s: State, left: Chunk[O], right: Chunk[O1]): (State, Chunk[Either[O, O1]]) = {
+      val leftsAvailable = left.size.toLong - s.leftDebt
+      if (leftsAvailable > 0) {
+        val leftWinners     = left.takeRight(leftsAvailable.toInt).map(Left(_))
+        val rightsAvailable = right.size.toLong - s.rightDebt - leftsAvailable
+        if (rightsAvailable > 0) {
+          val rightWinners = right.takeRight(rightsAvailable.toInt).map(Right(_))
+          (State(rightsAvailable, 0), leftWinners ++ rightWinners)
+        } else {
+          (State(0, -rightsAvailable), leftWinners)
+        }
+      } else {
+        val rightsAvailable = right.size - s.rightDebt
+        if (rightsAvailable > 0) {
+          val rightWinners = right.takeRight(rightsAvailable.toInt).map(Right(_))
+          (State(-leftsAvailable + rightsAvailable, 0), rightWinners)
+        } else {
+          (State(-leftsAvailable, -rightsAvailable), Chunk.empty)
+        }
+      }
+    }
+
+    ZTransducer {
+      for {
+        ref <- ZRef.make[State](State(0, 0)).toManaged_
+        p1  <- self.push
+        p2  <- that.push
+      } yield { (in: Option[Chunk[I1]]) =>
+        for {
+          st       <- ref.get
+          newState <- p1(in).zipWithPar(p2(in)) { case (lefts, rights) => updateState(st, lefts, rights) }
```

Comment:
This will both transducers in parallel and only emit when both of them are done. I think that's a bit misleading for a `race` operation, no?

Reply:
Yeah. Good catch. I should add a testcase with a never-emitting transducer.

Reply:
I added a test where one of transducers never emits and this test passes.
Here I wait for both sides to process a single chunk (resulting in potentially empty outputs).
It doesn't wait for both sides to actually emit a value.

Or do you mean we should go further and interrupt the last push of the slowest transducer?

Reply:
@simpadjo your test uses `filter` which emits empty chunks. Try using mapM with ZIO.never.

Reply:
Racing with such transducers can't work, as far as I understand. 
Transducers process the input sequentially. So if we don't want to wait for the result of `never`, me must stage incoming chunks. But this would lead to space leaks. 
Such optimization would work for the very last push tho. 
If you think that this is beneficial, I would be happy to explore.